buffer overrun in pipeline_bspatch_process

With a modern gcc 8.2.0 compilation fails because of:

```
src/core/pipeline/bspatch.c: In function ‘pipeline_bspatch_process’:
src/core/pipeline/bspatch.c:80:30: warning: array subscript 8 is above array bounds of ‘uint8_t[8]’ {aka ‘unsigned char[8]’} [-Warray-bounds]
                     newsize.c[j-16]= *bufp & 0x7F;
```

--- a/build/zephyr/application/CMakeLists.txt
+++ b/build/zephyr/application/CMakeLists.txt
@@ -1,5 +1,5 @@
 include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
-project(test_memory)
+project(application)
 
 target_sources(app PRIVATE src/runner.c)
 
@@ -11,7 +11,7 @@ file(GLOB libpull_headers ${ROOTDIR}/include)
 target_sources(app PRIVATE ${libpull_src})
 include_directories(${libpull_headers} ${ROOTDIR})
 
-add_definitions(-DSIMPLE_MANIFEST=1 -DLOGGER_VERBOSITY=4)
+add_definitions(-DSIMPLE_MANIFEST=1 -DLOGGER_VERBOSITY=1)
 add_definitions(-DMEMORY_FLASH_IMPL=1 -DWITH_TINYCRYPT=1)
 add_definitions(-DWITH_UPDATE_AGENT=1 -DWITH_CONNECTION_ZOAP=1)
 add_definitions(-DWITH_PLATFORM_NRF52840=1)

--- a/build/zephyr/application_ble/CMakeLists.txt
+++ b/build/zephyr/application_ble/CMakeLists.txt
@@ -1,13 +1,12 @@
 include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
-project(test_memory)
+project(application_ble)
 
-file(GLOB app_src src/*.c)
-target_sources(app PRIVATE ${app_src})
+ target_sources(app PRIVATE src/main.c)
 
 set(ROOTDIR "../../..")
 
 # Add libpull to the build
-file(GLOB libpull_src ${ROOTDIR}/src/**/*.c)
+file(GLOB_RECURSE libpull_src ${ROOTDIR}/src/*.c)
 file(GLOB libpull_headers ${ROOTDIR}/include)
 target_sources(app PRIVATE ${libpull_src})
 include_directories(${libpull_headers} ${ROOTDIR})
@@ -15,10 +14,7 @@ include_directories(${libpull_headers} ${ROOTDIR})
 add_definitions(-DSIMPLE_MANIFEST=1 -DLOGGER_VERBOSITY=4)
 add_definitions(-DMEMORY_FLASH_IMPL=1 -DWITH_TINYCRYPT=1)
 add_definitions(-DDISABLE_LIBPULL_NETWORK=1 -DBUFFERED_WRITER=1)
+add_definitions(-DWITH_ZEPHYR_BLE=1)
+add_definitions(-DWITH_PLATFORM_NRF52840=1)
 
-# Add the board files
-set(BOARDDIR "../board/${BOARD}")
-include_directories(${BOARDDIR})
-include_directories(${ROOTDIR}/test/platform)
-file(GLOB board_src ${BOARDDIR}/*.c)
-target_sources(app PRIVATE ${board_src})
+include_directories(${ROOTDIR}/src/platform/nrf52840)

--- a/build/zephyr/application_ble/src/main.c
+++ b/build/zephyr/application_ble/src/main.c
@@ -11,6 +11,12 @@
 #define DEVICE_NAME	CONFIG_BT_DEVICE_NAME
 #define DEVICE_NAME_LEN	(sizeof(DEVICE_NAME) - 1)
 
+#if  WITH_TINYCRYPT
+int default_CSPRNG(uint8_t *dest, unsigned int size) {
+    return 0;
+}
+#endif
+
 static const struct bt_data ad[] = {
 	BT_DATA_BYTES(BT_DATA_FLAGS, BT_LE_AD_NO_BREDR),
 	BT_DATA_BYTES(BT_DATA_UUID16_ALL, 0x0a, 0x18),
@@ -45,8 +51,7 @@ static struct bt_conn_cb conn_callbacks = {
 	.disconnected = disconnected,
 };
 
-int main(void)
-{
+int main(void) {
 	int err;
 
 	err = bt_enable(NULL);

--- a/build/zephyr/bootloader/CMakeLists.txt
+++ b/build/zephyr/bootloader/CMakeLists.txt
@@ -1,5 +1,5 @@
 include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
-project(test_memory)
+project(bootloader)
 
 target_sources(app PRIVATE src/bootloader.c)
 
@@ -11,7 +11,7 @@ file(GLOB libpull_headers ${ROOTDIR}/include)
 target_sources(app PRIVATE ${libpull_src})
 include_directories(${libpull_headers} ${ROOTDIR})
 
-add_definitions(-DSIMPLE_MANIFEST=1 -DLOGGER_VERBOSITY=4)
+add_definitions(-DSIMPLE_MANIFEST=1 -DLOGGER_VERBOSITY=1)
 add_definitions(-DMEMORY_FLASH_IMPL=1 -DDISABLE_LIBPULL_NETWORK=1)
 add_definitions(-DWITH_TINYCRYPT=1 -DWITH_BOOTLOADER=1)
 add_definitions(-DWITH_PLATFORM_NRF52840=1 -DWITH_ARM_CORTEX_LOADER=1)

--- a/src/core/pipeline/bspatch.c
+++ b/src/core/pipeline/bspatch.c
@@ -64,6 +64,10 @@ int pipeline_bspatch_process(pipeline_ctx_t* ctx, uint8_t* buf, int len) {
 	    oldpos=0;
         newpos=0;
 
+        // XXX This code can be refactored as a while loop with different cases.
+        // This will reduce the number of exit point from the Duff's device and
+        // make the code much smaller (around 1kB);
+        //
         // Get header and check size
         for (j=0; j<24; j++) {
             while ((buf+len-bufp) == 0) {
@@ -74,7 +78,7 @@ int pipeline_bspatch_process(pipeline_ctx_t* ctx, uint8_t* buf, int len) {
                     return -1;
                 }
             } else {
-                if (j != (16+8)) {
+                if (j != (16+7)) {
                     newsize.c[j-16] = *bufp;
                 } else {
                     newsize.c[j-16]= *bufp & 0x7F;

--- a/src/network/ble/zephyr_ble.c
+++ b/src/network/ble/zephyr_ble.c
@@ -9,10 +9,9 @@
 
 #include <libpull/common.h>
 #include <libpull/network/writer.h>
+#include <libpull/network/gatt.h>
 
 #include "platform_headers.h"
-
-#include "libpull_gatt.h"
 
 libpull_fsm_ctx_t ctx;
 

--- a/test/support/utils.c
+++ b/test/support/utils.c
@@ -6,12 +6,13 @@ int file_compare(char* fname1, char* fname2) {
     int ch1, ch2;
     fp1 = fopen(fname1, "r");
     if (fp1 == NULL) {
-        printf("Cannot open %s for reading\n", fname1);
+        fprintf(stderr, "Cannot open %s for reading\n", fname1);
         return 1;
     }
     fp2 = fopen(fname2, "r");
     if (fp2 == NULL) {
-        printf("Cannot open %s for reading\n", fname2);
+        fclose(fp1);
+        fprintf(stderr, "Cannot open %s for reading\n", fname2);
         return 1;
     }
     do {

--- a/utils/assets_generator.sh
+++ b/utils/assets_generator.sh
@@ -24,8 +24,9 @@ fi
 echo "Generate Test Memory file..."
 test_memory_file=$ASSETSDIR/test_memory
 max=64
-for i in $(seq 1 $max); do
+while [ $max -gt 0 ]; do
     echo "memory_test" >> $test_memory_file
+    (( max-- ))
 done
 echo "Generate Test Memory file...done"
 

--- a/utils/firmware_tool/commands/manifest.cpp
+++ b/utils/firmware_tool/commands/manifest.cpp
@@ -90,20 +90,24 @@ int manifest_generate_command(Context ctx) {
     uint8_t* signature = new uint8_t[64]; // XXX hardcoded value
     std::ifstream vend_priv_key(ctx.get_vendor_priv_key(), std::ios_base::binary);
     if (!vend_priv_key) {
+        delete[] vend_priv_key_buffer;
+        delete[] signature;
         std::cout << "Impossible to read the vendor private key" << std::endl;
         return EXIT_FAILURE;
     }
     vend_priv_key.read((char*)vend_priv_key_buffer, 32); // TODO check for errors && XXX hardcoded value
     vend_priv_key.close();
     err = sign_manifest_vendor(&mt, digest, vend_priv_key_buffer, signature, tinycrypt_secp256r1_ecc);
+    delete[] vend_priv_key_buffer;
     if (err) {
+        delete[] signature;
         std::cout << "Error signing the manifest" << std::endl;
         return EXIT_FAILURE;
     }
-    delete[] vend_priv_key_buffer;
     // (7) Store the signature in the metadata and a standalone file
     std::ofstream out_signature(ctx.get_signature_file(),  std::ios_base::binary);
     if (!out_signature) {
+        delete[] signature;
         std::cout << "Error writing on file: " << ctx.get_signature_file() << std::endl;
         return EXIT_FAILURE;
     }
@@ -116,18 +120,19 @@ int manifest_generate_command(Context ctx) {
     uint8_t* server_priv_key_buffer = new uint8_t[32];
     std::ifstream server_priv_key(ctx.get_server_priv_key(), std::ios_base::binary);
     if (!server_priv_key) {
+        delete[] server_priv_key_buffer;
         std::cout << "Impossible to read the server private key" << std::endl;
         return EXIT_FAILURE;
     }
     server_priv_key.read((char*)server_priv_key_buffer, 32); // TODO check for errors && XXX hardcoded value
     server_priv_key.close();
     err = sign_manifest_server(&mt, digest, server_priv_key_buffer, signature, tinycrypt_secp256r1_ecc);
+    delete[] server_priv_key_buffer;
+    delete[] signature;
     if (err) {
         std::cout << "Error signing the manifest" << std::endl;
         return EXIT_FAILURE;
     }
-    delete[] signature;
-    delete[] server_priv_key_buffer;
     // (9) write the metadata to the designed binary file
     std::ofstream out(ctx.get_out_file(), std::ios_base::out | std::ios_base::binary);
     if (!out) {

--- a/utils/firmware_tool/commands/pipeline.cpp
+++ b/utils/firmware_tool/commands/pipeline.cpp
@@ -798,6 +798,7 @@ int pipeline_patch_command(Context ctx) {
         input1.close();
         input2.close();
         outfile.close();
+        delete[] buffer;
         return -1;
     }
     // 6. Store the patched file

--- a/utils/firmware_tool/commands/pipeline.cpp
+++ b/utils/firmware_tool/commands/pipeline.cpp
@@ -169,8 +169,8 @@ int pipeline_compress_command(Context ctx) {
         }
     }
     flush_bit_buffer(outfile);
-    printf("text:  %ld bytes\n", textcount);
-    printf("code:  %ld bytes (%ld%%)\n",
+    printf("text:  %lu bytes\n", textcount);
+    printf("code:  %lu bytes (%lu%%)\n",
             codecount, (codecount * 100) / textcount);
     return 0;
 }

